awx/main/models/workflow.py

```diff
@@ -682,6 +681,27 @@ def get_ancestor_workflows(self):
             wj = wj.get_workflow_job()
         return ancestors
 
+    def get_effective_artifacts(self, parents_set=None):
+        """
+        For downstream jobs of a workflow nested inside of a workflow,
+        we send aggregated artifacts from the nodes inside of the nested workflow
+        """
+        artifacts = {}
+        job_queryset = (
+            UnifiedJob.objects.filter(unified_job_node__workflow_job=self)
+            .defer('job_args', 'job_cwd', 'start_args', 'result_traceback')
+            .order_by('status', 'finished', 'id')
+            .iterator()
+        )
+        if parents_set is None:
+            parents_set = set()
+        new_parents_set = parents_set | set([self.id])
```

Comment:
I'd like to start getting away from Python 2-isms like `set([...])`.

Reply:
Is it even possible to have a cycle in the jobs?  The job templates, yes, I could see that potentially happening, but even if the same JT is used in multiple places won't it spawn distinct new jobs (thus having new pks)?

Reply:
No, as structured I don't think it would have any cycles. The map of workflow jobs would always be a tree, since a workflow job node can't (in practice) link to the same workflow job as another job.

I kept coming back to questioning if we should track and skip duplicate _workflow job templates_, which is a much stronger criteria. A workflow can use the same WFJT inside of it multiple times, or it can be nested deeper. This would cut down on queries, but I'm not sure if users might actually want to combine them because multiple runs of the same WFJT could give different artifacts.